Prevent instability and crash due to message flood

tu-maurice opened the pull request:
This patch adds a rate limiting to selected patches. The underlying rate limiter
used is the Leaky-Bucket algorithm. It allows for a burst of messages, but
limits them after a specified amount of messages within a time frame.
If the ratelimit hits the messages are simply ignored.

For now its set to a burst of 30 allowed messages and a subsequent limit of 4 messages per second, which seems to do the trick without inconveniencing normal users.

It should to some extend prevent the recent issues with bots decribed in #3505.

What was tested with the patch:

- Joining (works)
- Talking with UDP and TCP mode (works)
- Switching channels (works)
- Bringing down the server with a bot, tested with two different bots (does not work)

I'm thanking the Zom.bi community for testing and @Natenom for some insight into the problem.

I'm open to any suggestions.

tu-maurice commented:
@davidebeatrici May I ask what priority/P0 entails?

davidebeatrici commented:
From https://wiki.mumble.info/wiki/Issue_Priorities:
> MUST be fixed ASAP. Issues that prevent further development; legal issues; serious data loss issues; build issues for platforms we provide binaries for.

tu-maurice commented:
@davidebeatrici How may I understand this? Is it my patch that "MUST be fixed ASAP" or does the patch fix something that "MUST be fixed ASAP"?

davidebeatrici commented:
The patch fixes something that "MUST be fixed ASAP".

davidebeatrici commented:
Qt 4 builds failing because:
```c++
In file included from /usr/include/c++/4.8/chrono:35:0,
                 from ./ServerUser.h:17,
                 from ../ACL.cpp:14:
/usr/include/c++/4.8/bits/c++0x_warning.h:32:2: error: #error This file requires compiler and library support for the ISO C++ 2011 standard. This support is currently experimental, and must be enabled with the -std=c++11 or -std=gnu++11 compiler options.
 #error This file requires compiler and library support for the \
  ^
```

MinGW builds failing because:
```c++
ServerUser.h:106:36: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
   bool limit = this->currentTokens > (MAX_TOKENS - tokens);
                                    ^
```

tu-maurice commented:
Alright, I can work with that. The first problem should be fixable by relying on pre-C++11 means to measure time if the software is not build using at least a C++11 compiler. The second problem I didn't see, but is easily fixable.

Kissaki commented:
After landing we should also put this into 1.2.

davidebeatrici commented:
I think that we shouldn't create another 1.2.x release.

tu-maurice commented:
This patch should allow for pre-C++11 builds and fixes the warning with the unsigned signed comparison.

tu-maurice commented:
I hope you don't mind that I squashed them, so the history looks nicer. :+1:

davidebeatrici commented:
No problem, we usually prefer commits to be squashed if they're all related.

tu-maurice commented:
Perfect. Now do the last changes solve the problems you mentioned early?

tu-maurice commented on src/murmur/ServerUser.cpp:
```diff
@@ -112,3 +112,53 @@ int BandwidthRecord::bandwidth() const {
 }
 
+// Rate limiting: burst up to 30, 4 message per sec limit over longer time
+LeakyBucket::LeakyBucket() : tokensPerSec(4), maxTokens(30), currentTokens(0) {
```
If you keep tokensPerSec and maxTokens constant then I don't see why these should be members. Maybe add them as parameters in the constructor too, so it can be set from for example the murmur.ini.

Or if you just don't want the ServerUser.h to be cluttered then move the LeakyBucket stuff to an extra header file.

davidebeatrici replied:
I set the values in the constructor in preparation for a pull request which adds the two parameters in the server configuration (`murmur.ini`).

davidebeatrici commented:
Builds succeeded.

Feel free to review my changes and squash the commits.

tu-maurice commented:
I separated the now and millisecondsBetween methods from the LeakyBucket class, as they were neither operating directly on the object nor public. And I added static inline to each to encourage the compile to inline them.

Also I moved a comment.

Please look at it again to see if you're okay with that.

davidebeatrici commented:
Thank you very much for your contribution!
